test/Concurrency/Runtime/async_custom_task_executor.swift

```diff
@@ -0,0 +1,99 @@
+// RUN: %target-run-simple-swift( -Xfrontend -disable-availability-checking -parse-as-library )
+
+// REQUIRES: executable_test
+// REQUIRES: concurrency
+// REQUIRES: libdispatch
+
+// REQUIRES: concurrency_runtime
+// UNSUPPORTED: back_deployment_runtime
+
+import Dispatch
+import StdlibUnittest
+
+final class CountEnqueuesSerialExecutor: SerialExecutor, @unchecked Sendable {
+  var enqueueCount = 0
+
+  func enqueue(_ job: UnownedJob) {
+    enqueueCount += 1
+    print("enqueue job")
+  }
+}
+
+final class CountEnqueuesExecutor: Executor, @unchecked Sendable {
+  var enqueueCount = 0
+
+  func enqueue(_ job: UnownedJob) {
+    enqueueCount += 1
+    print("enqueue job")
+  }
+}
+
+
+@MainActor
+func mainActorMethod() {}
+
+@available(SwiftStdlib 5.9, *)
+actor CounterExecutorActor {
+  let executor: CountEnqueuesSerialExecutor
+
+  nonisolated var unownedExecutor: UnownedSerialExecutor {
+    self.executor.asUnownedSerialExecutor()
+  }
+
+  init(executor: CountEnqueuesSerialExecutor) {
+    self.executor = executor
+  }
+
+  func test() {
+    fatalError("Not expecting to run this")
+  }
+}
+
+@globalActor
+public actor MyGlobalActor: GlobalActor {
+  public static let shared = MyGlobalActor()
+
+  let executor = CountEnqueuesSerialExecutor()
+
+  @inlinable
+  public nonisolated var unownedExecutor: UnownedSerialExecutor {
+    self.executor.asUnownedSerialExecutor()
+  }
+}
+
+@main struct Main {
+
+  static func main() {
+    if #available(SwiftStdlib 5.9, *) {
+      let countingExecutor = CountEnqueuesSerialExecutor()
+
+      Task(on: countingExecutor) {
+        fatalError("never executed")
+      }
+      assert(countingExecutor.enqueueCount == 1)
+
+      let actor = CounterExecutorActor(executor: countingExecutor)
+      Task(on: actor) { actor in
+        actor.test()
+        fatalError("never executed")
+      }
+      assert(countingExecutor.enqueueCount == 2)
+
+      Task(on: MainActor.self) { // specialized method for main actor
+        mainActorMethod() // good, no await here
+      }
+      assert(countingExecutor.enqueueCount == 2)
+
+      MainActor.task { actor in
+        await mainActorMethod() // unfortunate await here, we are on the main actor after all
+      }
+      assert(countingExecutor.enqueueCount == 2)
+
+      assert(MyGlobalActor.shared.executor.enqueueCount == 0)
+      MyGlobalActor.task { actor in
+        // do nothing
+      }
+      assert(MyGlobalActor.shared.executor.enqueueCount == 1)
+    }
```

Comment:
^^^ Current list of APIs; the global actor ones we'll have to figure out better and it'd be nice to avoid the `actor in` dance for actors.